### ISSUE TITLE
Add some API to support GUI customization

### ISSUE THREAD

**Issue by NoNameProvided** [Microsoft/vscode#1833]
VS Code is a great coding IDE, but developers love to customize their tools to make them feel like home and VS Code lacks those customization abilities. 

I love the sublime material theme and I wanted to implement it in VS Code but as far as I can see it's not possible now because only the coding area is customizable. So in my opinion some improvement in ui customization would be welcomed by a lot of us. In the ideal word every pixel and element of VS Code should be customizable with extensions, but for start it would be enough to start with the following things:

 - customization of the side panel (#427 and #123)
 - adding support for custom tabs, the possibilities of the quick box is very limited (#710)
 - adding the ability to add and remove ui elements (e.g. a bottom panel for error listing like in Atom)

I have read about the [approach](https://code.visualstudio.com/docs/extensions/our-approach) followed by VS Code. So instead of allowing extensions to create custom ui elements, VS Code itself should provide some API for this, what allows to the developers to set the content, but the styling of the content should be based on the theme of VS Code.

So in a nutshell: VS Code core provide the bones and template of the ui elements, extension developers can set the content but not the style and theme developers can style the content.

**Comment by zamamohammed** [comment]
:+1:

**Comment by joaomoreno** [comment]
@egamma @chrisdias fyi

**Comment by egamma** [comment]
see also #1231

**Comment by siegebell** [comment]
I wrote about this a bit in #272.

At the moment, you can write a custom UI by showing a markdown document in preview mode that contains an iframe to an HTML file that communicates with your extension via web sockets. So far, it has worked really well for me. It will get a bit easier when the virtual read-only document feature is released (#272). But it cannot integrate with the rest of the vscode UI and getting the right theme is tricky (you can access vscode's markdown css by recursively looking up the iframe parent's style sheet).

(For those interested, you can see the HTML view I wrote [here](https://github.com/siegebell/vscoq/tree/master/client/src/HtmlView) and the code that connects the HTML view to vscode [here](https://github.com/siegebell/vscoq/blob/master/client/src/HtmlCoqView.ts)).

But the response from @jrieken was:
> the preview mechanism isn't designed to be a UI extension API. While you can render html and run scripts you should not go crazy in there.

So it's not clear to me whether this will remain a viable/supported solution in the future.

In particular, I am unsure whether the following comment,
> The updating (when a document emits a change event) is not built/tested for high frequency of changes,

means that high-frequency changes are not intended for 1) the HTML via the virtual document text changes or 2) via the javascript running in the virtual document preview modifying the DOM.

Could we have some clarification here? Is a custom UI via HTML, javascript, and web-sockets a dead-end solution (that will break in the future)? Will an API be provided in the future or is it an intended limitation? Are the devs willing to support hosting a custom UI in the preview as a stop-gap solution until a proper API is provided?

I can accept that the vscode devs are too busy to implement or even specify such an API right now. But some guidance here would help because we extension developers are impatient and willing to rely on hacks.

**Comment by jrieken** [comment]
First: What you are doing is pretty cool - sending out some html and talking back via web sockets to your extension is a clever solution. I am glad that virtual documents make your life a little easier, btw even wrt to [styles](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/html/browser/htmlPreviewPart.ts#L182)

The intended use-case for virtual documents and the previewHtml-command is something like 'showing source from a dll-file' or 'generating html from markdown'. The assumption is that those things aren't huge and that they don't update like crazy. 

When it comes to UI we try to be smart - our tree and editor widgets are fully virtualised, render with the screen-refresh-rate, use diff'ing smartness etc. That's all for the sake of 'snappy VS Code'. The html preview part doesn't employ such techniques. And since (unlike extensions) it runs in the same process there is a chance of it wrecking VS Code and that's not what we want. 

So when it comes to updating a virtual document at a high rate you should not use the `onDidChange` [event](https://github.com/Microsoft/vscode/blob/master/src/vs/vscode.d.ts#L1059) because that's not [cleverly](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/html/browser/htmlPreviewPart.ts#L149) implemented (yet?).

Keep doing the smart thing you already do - we won't break you in the foreseeable future but it also depends on how well extensions doing this behave. If things like this cause VS Code to stall we have to react (at least blame the extension ;-))

**Comment by siegebell** [comment]
Once Chrome is done implementing OOPIFs ([issue 99379](https://code.google.com/p/chromium/issues/detail?id=99379) and [spec](http://www.chromium.org/developers/design-documents/oop-iframes)) there could be a way to run extension UIs in a separate process. Maybe in the same process as the extension, allowing it direct access to the DOM? Hmm...

**Comment by iDev0urer** [comment]
This would be a super nice feature to have! I would love to be able to modify the document tree a little bit so as to filter out some files that are junking up my view but I have to keep.

**Comment by bgashler1** [comment]
Related #459

**Comment by nulled666** [comment]
I would like to have some UI APIs that could help me create popup panels for css value tuning.

**Comment by be5invis** [comment]
A dirty hack to do this: https://github.com/be5invis/vscode-custom-css
Waiting official API. :(

**Comment by atian25** [comment]
+1

I have a similar feature request too:  how to custom sort order of sidebar files?

**Comment by trubit** [comment]
:+1:

**Comment by redVi** [comment]
I like VS Code really. But I want see customizable and beautiful editor. Right now it looks like one of the tools of Windows. I have OS X and the interface of this editor should not dramatically different from the native interface.

<img width="1275" alt="vs_code_ui" src="https://cloud.githubusercontent.com/assets/2623486/19519041/4b515aec-9634-11e6-89a1-0f9b08cfa1fd.png">

**Comment by Revod** [comment]
Any update about this request?

**Comment by be5invis** [comment]
@rebornix @joaomoreno any progress on this?

**Comment by Hendrixer** [comment]
like @be5invis  said, https://github.com/be5invis/vscode-custom-css works pretty great. I was able to achieve this in just 20 mins
<img width="1436" alt="screen shot 2016-11-18 at 2 46 43 pm" src="https://cloud.githubusercontent.com/assets/2840504/20444168/f88ce55c-ad9d-11e6-8f5d-34ec021636b8.png">

**Comment by chixu** [comment]
+1

**Comment by mszczepaniak** [comment]
@Hendrixer it looks great! Could You share Your config to achieve this colors?

**Comment by Revod** [comment]
@mszczepaniak @Hendrixer I want that config too please

**Comment by shaunlebron** [comment]
To be extra clear, VS Code does not support what Atom calls "UI Theme" below:

<img width="792" alt="no-ui-theme-vscode" src="https://cloud.githubusercontent.com/assets/116838/21692080/d85936b4-d340-11e6-8255-52e22eeecf14.png">

This can be jarring since syntax themes can clash with the surrounding UI design.

**Comment by bgashler1** [comment]
@shaunlebron we are going to support this soon #17608

**Comment by kabirbaidhya** [comment]
vscode is a great code editor with a lot of features and the hackability it provides the only thing it doesn't have is a good UI and look & feel. It think it's really important for vscode to provide the ability to customize the UI so that the community can come up with some better themes.

**Comment by freMea** [comment]
I think it is a duplicate of this [Support to theme the entire workbench #3112](https://github.com/Microsoft/vscode/issues/3112) issue.

**Comment by NoNameProvided** [comment]
I would say #3112 is the duplicate of this as 1833 < 3112

**Comment by orta** [comment]
Looks like out of process iframes (OOPIFs) are being [shipped](https://groups.google.com/a/chromium.org/forum/#!topic/site-isolation-dev/Nttn3WiY9DA)  [in Chrome](https://groups.google.com/a/chromium.org/forum/#!topic/chromium-dev/qlI54-dK4Ac) and so it may be available to be exposed in electron

**Comment by rei-vilo** [comment]
Any plans to improve the GUI features of Visual Studio Code? Last [release 1.17](http://blog.atom.io/2017/05/16/atom-1-17.html) of **Atom** features docks.

![docks-demo](https://cloud.githubusercontent.com/assets/951776/26157570/db5170e8-3b19-11e7-888f-7fa994f3a1f2.gif)

**Comment by xshotD** [comment]
Any updates on developments for this? Last update by devs for this was 4 months ago and we haven't seen anything happen since.

**Comment by marcoms** [comment]
It looks to me like VSCode isn't planning to be as flexible as Atom in this regard, instead prioritizing surface level adjustments like colors, fonts, etc. This will probably be better for performance, but personally I prefer Atom's tradeoff.

**Comment by RobertBColton** [comment]
+1 #5604 is related too. So many more extensions would be possible if UI customization was more robust. I need the ability to create custom editors with checkboxes, dropdowns, and form fields to edit metadata files for a game engine. Since an image editor that can preview images instead of showing them as blobs in the text editor, it should be possible to create more of these kinds of editors as plugins.

**Comment by cwtuan** [comment]
Any update?

**Comment by HarisHashim** [comment]
Is @marcoms comment on Jun 21 means that this issue is dead. Ie. the VS Code team will not provide such extension capability of Atoms as commented by @rei-vilo .

If so, will be sad indeed!

**Comment by orta** [comment]
If you go through the chrome issues, out of process iframes have been in
development for a while and seem to be wrapping up. So figuring out how
that could work with electron and then vscode would probably be useful for
this feature.

**Comment by rei-vilo** [comment]
The lack of GUI customisation, and more specifically the lack of support for a tool-bar, made me remove Visual Code Studio from my short-list.

![unknown](https://user-images.githubusercontent.com/951776/31315122-10d8e412-ac12-11e7-9e78-020c814746ec.png)

Visual Code Studio is based on [Electron](https://electron.atom.io/docs/), and Electron has a large offer of tool-bars —see [tool-bar](https://atom.io/packages/tool-bar), [Flex Tool Bar](https://atom.io/packages/flex-tool-bar), [Tool Bar Main](https://atom.io/packages/tool-bar-main), [Toolbar Almighty](https://atom.io/packages/tool-bar-almighty), [Toolbar for Atom](https://atom.io/packages/tool-bar-atom), [Toolbar Shortcuts](https://atom.io/packages/tool-bar-shortcuts), and [many more](https://atom.io/packages/search?utf8=%E2%9C%93&amp;q=keyword%3Atool-bar).

So, to rephrase the initial question, how to add an Electron plug-in to Visual Code Studio?

**Comment by marcoms** [comment]
@rei-vilo, those plugins are designed for Atom, and do not work universally across Electron apps

**Comment by watzon** [comment]
@rei-vilo expanding on @marcoms comment, I think you are confusing Electron with Atom. Electron is a javascript framework that allows you to make chromium based desktop applications using javascript and HTML. Atom is an open source text editor built by the folks at GitHub that was built using the Electron framework.

**Comment by rei-vilo** [comment]
Ok, thanks!

**Comment by mrpotatoes** [comment]
This is really unfortunate as when I started to use VS Code this week I was actually really excited about it's system but if I can't customize it's GUI much if at all then I don't see the point in using this over Atom.

**Comment by watzon** [comment]
@mrpotatoes VS Code is newer than Atom still so give it some time to catch up. It's also faster by a lot and haing plugins written in Typescript by default it awesome

**Comment by flyisland** [comment]
just took a look at the [roadmap of 2018](https://github.com/Microsoft/vscode/wiki/Roadmap), it seems like the team still have no plan for this request on next year, too bad!

**Comment by rei-vilo** [comment]
Stefano:

It is currently 15:42 CET = 14:42 GMT, so the conference starts within 18 minutes at 16:00 CET = 15:00 GMT.

I had to lobby hard to obtain GMT hour to be mentioned.

See you then!


— Olivier
rei_vilo@hotmail.com<mailto:rei_vilo@hotmail.com>


Le 20 déc. 2017 à 15:40, flyisland <notifications@github.com<mailto:notifications@github.com>> a écrit :


just took a look at the roadmap of 2018<https://github.com/Microsoft/vscode/wiki/Roadmap>, it seems like the team still have no plan for this request on next year, too bad!

**Comment by tracker1** [comment]
Was thinking that the way plugins run with VS Code, maybe exposing a render interface, similar to React's would be beneficial... this way the render method will return a virtual dom representation in the extension/plugin, and the interface then swaps out any attached methods for events with dispatchers that will dispatch through to the extension, and trigger that a fresh render is needed.

Even if a plugin's render takes too long, it won't hang the rest of the application...

it would be like react, but split across a boundary where the composition is in the plugin/extension, and the actual render-diff happens in vs code's UI.

**Comment by janaka** [comment]
Here's the use case I had in mind. Recently I've started contributing to the .NET Test Explorer extension. As far as I know it's the only extension available for working with unit tests in a .NET Core project. Having a powerful scalable (large number of tests) tool in this regard is important if VS Code is to be successful as the IDE for .NET Core development on MacOS. It's to have at least some of the testing power features as VS.

The explorer and command palette UI options are too limited for building such experiences. Currently .NET Test Explorer renders as an Explorer Tree view. It occurred to me earlier that a UI similar to the Search option in VS Code could be a good fit. Allowing a developer to quickly free text search for tests, filter in failing tests, and re-run just those tests with a keyboard shortcut would be an example workflow.

As far as I can tell there are no facilities at present to build such an experience as an extension.

**Comment by mrpotatoes** [comment]
[It's on the docket!](https://github.com/Microsoft/vscode/issues/43361)
> Investigate promoting explorer contributions to viewlets #43645 @sandy081

(I know I said I couldn't use this over Atom but in the past month I've been using it over Atom so I'm interested in it's progress again. Forgive me)

**Comment by simon-robertson** [comment]
Personally, the "default dark" theme is fine for me. The only related request I have is to allow the status bar to have a "dark" theme as well, to match the rest of the GUI; the default blue theme is too distracting.

**Comment by rei-vilo** [comment]
![toolbar](https://user-images.githubusercontent.com/951776/51441776-67682100-1cd5-11e9-8dbe-9edf705f6500.png)

> Tip: Use the setting `debug.toolBarLocation to control the location of the debug toolbar. It can either be the default `floating`, `docked` to the debug viewlet or hidden`. A `floating` debug toolbar can be dragged horizontally and also down to the editor area.


as per the [Debugging](https://code.visualstudio.com/Docs/editor/debugging) page.

**Comment by rei-vilo** [comment]
It looks like [JerryGoyal](https://github.com/JerryGoyal) found a solution with the [Shortcut Menu Bar VSCode Extension](https://github.com/JerryGoyal/Shortcut-Menu-Bar-VSCode-Extension).

<img width="246" alt="capture 2019-02-08 a 20 23 00" src="https://user-images.githubusercontent.com/951776/52501685-364a8480-2be1-11e9-94b1-d49484948753.png">

**Comment by tristan957** [comment]
I am going to throw my hat in the ring as well. When Firefox supported GUI customization through XUL, I could permanently hide the horizontal tabs at the top. I would love to do this in VSCode. I think vertical tabs work better in some cases especially on wide screens.

**Comment by madushajg** [comment]
bump

**Comment by probablykasper** [comment]
@jrieken Sorry to respond to this literally 4 years later, but you explained that allowing this could have cause VSCode itself to get slower or freeze. Maybe it's worth looking into how Atom's implementation of this and evaluate that?

One solution would be to give users a warning when installing extensions using this feature. That sounds like an alright compromise for me, to get an important feature as this.

**Comment by zvidan** [comment]
Hi @rei-vilo
You have mentioned the Debug toolbar above:
https://github.com/microsoft/vscode/issues/1833#issuecomment-455878771
Do you know, if some ability to change it (add button) does exist ?
In my opinion, "Show Next Statement" (as in classic Visual Studio) is mandatory here !

**Comment by rei-vilo** [comment]
@zvidan It has been a very long time since my comment. I suggest to open a new issue.
